nvtabular/io.py

```diff
@@ -228,55 +235,68 @@ def __init__(
         # Resolve file system
         self.fs = fs or get_fs_token_paths(str(out_dir))[0]
 
-        # create thread queue and locks
-        self.queue = queue.Queue(num_threads)
-        self.write_locks = [threading.Lock() for _ in range(num_out_files)]
+        # Only use threading if num_threads > 0
+        if self.num_threads:
 
-        # signifies that end-of-data and that the thread should shut down
-        self._eod = object()
+            # create thread queue and locks
+            self.queue = queue.Queue(num_threads)
+            self.write_locks = [threading.Lock() for _ in range(num_out_files)]
 
-        # create and start threads
-        for _ in range(num_threads):
-            write_thread = threading.Thread(target=self._write_thread, daemon=True)
-            write_thread.start()
+            # signifies that end-of-data and that the thread should shut down
+            self._eod = object()
+
+            # create and start threads
+            for _ in range(num_threads):
+                write_thread = threading.Thread(target=self._write_thread, daemon=True)
+                write_thread.start()
 
     def set_col_names(self, labels, cats, conts):
         self.cats = cats
         self.conts = conts
         self.labels = labels
         self.column_names = labels + conts
 
+    def _write_table(self, idx, data):
+        return
+
     def _write_thread(self):
         return
 
     @annotate("add_data", color="orange", domain="nvt_python")
     def add_data(self, gdf):
 
-        # Shuffle if necessary
-        # (Skip shuffle if "full", because we will do it later)
-        if self.shuffle and self.shuffle != "full":
-            gdf = _shuffle_gdf(gdf)
-
         # Populate columns idxs
         if not self.col_idx:
             for i, x in enumerate(gdf.columns.values):
                 self.col_idx[str(x)] = i
 
-        # get slice info
-        int_slice_size = gdf.shape[0] // self.num_out_files
-        slice_size = int_slice_size if gdf.shape[0] % int_slice_size == 0 else int_slice_size + 1
-
-        for x in range(self.num_out_files):
-            start = x * slice_size
-            end = start + slice_size
-            # check if end is over length
-            end = end if end <= gdf.shape[0] else gdf.shape[0]
-            to_write = gdf.iloc[start:end]
-            self.num_samples[x] = self.num_samples[x] + to_write.shape[0]
-            self.queue.put((x, to_write))
+        # Generate `ind` array to map each row to an output file.
+        # This approach is certainly more optimized for shuffling
+        # than it is for non-shuffling, but using a single code
+        # path is probably worth the (possible) minor overhead.
+        nrows = gdf.shape[0]
+        typ = np.min_scalar_type(nrows * 2)
+        if self.shuffle and self.shuffle != "full":
+            ind = cp.random.choice(np.arange(self.num_out_files, dtype=typ), nrows)
```

Comment:
Can we generate the range on the device (like cp.arange instead of np.arange)?  
```suggestion
            ind = cp.random.choice(cp.arange(self.num_out_files, dtype=typ), nrows)
```

Reply:
oops typo